relaynet-cogrpc-jvm/src/main/java/tech/relaycorp/relaynet/cogrpc/client/CogRPCClient.kt

```diff
@@ -1,19 +1,132 @@
 package tech.relaycorp.relaynet.cogrpc.client
 
+import io.grpc.netty.NettyChannelBuilder
+import io.grpc.stub.StreamObserver
+import kotlinx.coroutines.channels.BroadcastChannel
+import kotlinx.coroutines.channels.sendBlocking
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onEach
+import tech.relaycorp.relaynet.cogrpc.Authorization
+import tech.relaycorp.relaynet.cogrpc.CargoDelivery
+import tech.relaycorp.relaynet.cogrpc.CargoDeliveryAck
+import tech.relaycorp.relaynet.cogrpc.CargoRelayGrpc
 import tech.relaycorp.relaynet.cogrpc.CogRPC
+import java.net.InetSocketAddress
+import java.net.URL
+import java.util.logging.Level
+import java.util.logging.Logger
 
-interface CogRPCClient {
+class CogRPCClient(
+    serverAddress: String,
+    useTls: Boolean = true
+) {
+
+    private val channel by lazy {
+        val url = URL(serverAddress)
+        val address = InetSocketAddress(
+            url.host,
+            url.port.let { if (it != -1) it else DEFAULT_PORT }
+        )
+        NettyChannelBuilder
+            .forAddress(address)
+            .useTransportSecurity()
+            .run { if (useTls) useTransportSecurity() else usePlaintext() }
+            .build()
+    }
 
     @Throws(Exception::class)
-    fun deliverCargo(cargoes: Iterable<CogRPC.MessageDelivery>): Flow<CogRPC.MessageDeliveryAck>
+    fun deliverCargo(cargoes: Iterable<CogRPC.MessageDelivery>): Flow<CogRPC.MessageDeliveryAck> {
+        if (cargoes.none()) return emptyFlow()
+
+        val cargoesToAck = cargoes.map { it.localId }.toMutableList()
```

Comment:
That wouldn't work the way I was doing it. Because I was relying on having the full list to know when to close the ackStream. But now I've refactored it to wait for the delivery stream to complete.